### PR TITLE
[diagnostic-verifier] Add LLVM Options so that Diagnostic Verifier aborts when it emits error diagnostics

### PR DESCRIPTION
**Pull request by gottesmm**
Otherwise, it can be hard to figure out why the diagnostic verifier is actually failing. This change does this by:

1. Splitting up the huge method DiagnosticVerifier::verifyFile()[~632 lines of code in one method!] into several subroutines that have semantic names that tell big picture what the code is trying to do there.

2. Then I added `llvm::cl::opt<bool>` options that a compiler developer can use to cause the verifier to abort at the beginning of each of these helper functions. This lets one easily drop down into the debugger at the point where the "incorrect diagnostic" or "unexpected diagnostic" error occurs.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
Oh I mentioned in the commit msgs, but not in the PR msg. I did the refactoring using Xcode's refactoring engine!

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 1]
```diff
@@ -601,10 +651,23 @@ DiagnosticVerifier::Result DiagnosticVerifier::verifyFile(unsigned BufferID) {
...
-  
-  // Make sure all the expected diagnostics appeared.
-  std::reverse(ExpectedDiagnostics.begin(), ExpectedDiagnostics.end());
+void DiagnosticVerifier::verifyAllExpectedDiagnosticsAppeared(
+    StringRef BufferName, StringRef InputFile,
+    std::vector<ExpectedDiagnosticInfo> &ExpectedDiagnostics,
+    std::vector<llvm::SMDiagnostic> &Errors) {
+  using llvm::SMLoc;
+
+  auto addError = [&](const char *Loc, const Twine &message,
```
Comment (brentdax) [inline lib/Frontend/DiagnosticVerifier.cpp:662]: Can the various `addError` lambdas be a single, shared function or method? It looks like the only difference is which boolean you're testing.

**Review by beccadax — CHANGES_REQUESTED** [review]
One optional suggestion, but this otherwise looks okay to me.

File: lib/Frontend/DiagnosticVerifier.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/Frontend/DiagnosticVerifier.cpp:662]: Hmm... I think I have something I can do... Let me see.

**Comment by gottesmm** [comment]
I think I already did this but in a different PR.

**Pull request closed without merging**